enhancements/update/cluster-version-operator-x509-upstream-trust.md

```diff
@@ -0,0 +1,202 @@
+---
+title: cluster-version-operator-x.509-upstream-trust
+authors:
+  - "@wking"
+reviewers:
+  - "@LalatenduMohanty"
+approvers:
+  - "@bparees"
+  - "@deads2k"
+creation-date: 2020-05-12
+last-updated: 2020-05-14
+status: provisional
+see-also:
+  - "/enhancements/automated-service-ca-rotation.md"
+  - "/enhancements/kube-apiserver/certificates.md"
+  - "/enhancements/kube-apiserver/tls-config.md"
+  - "/enhancements/proxy/global-cluster-egress-proxy.md"
+  - "https://github.com/openshift/enhancements/pull/115"
+---
+
+# Cluster-version Operator X.509 Upstream Trust
+
+This enhancement provides a mechanism for providing [the cluster-version operator][cluster-version-operator] (CVO) with an alternative trust-store and TLS configuration to use when connecting to [the update recommendation service][update-service] that overrides both the CVO's defaults and any [proxy configuration](../proxy/global-cluster-egress-proxy.md).
```

Comment:
why would it override the proxy configuration?  if the proxy config is applicable (the host isn't in no_proxy) it should be used

Reply:
> if the proxy config is applicable (the host isn't in no_proxy) it should be used

This endpoint-level config seems more specific than the cluster-scoped Proxy config.  If the user wants the Proxy config to apply, they can unset this config, right?  But if the Proxy config took precedence, there would be no chance for the user to override for this specific endpoint.

Reply:
if they want the endpoint specific CAs to take precedence, they should be listed in "noproxy".  If the connection is going through the proxy, then the proxy CAs need to be used.

Reply:
>  If the connection is going through the proxy, then the proxy CAs need to be used.

How do we determine this, in a world where transparent, network-defined proxies may be set?  Say you are in a cluster where `upstreamTLS.trustedCA` is set in ClusterVersion and Proxy has `trustedCA` but neither `httpProxy` nor `httpsProxy`.  Which CA would you use connecting to an HTTPS `upstream`?  I'd expect `upstreamTLS.trustedCA`, because why would the admin have set that unless they expected us to use it?  Are you saying that, in the absence of `httpProxy` (for HTTP connections) or `httpsProxy` (for HTTPS connections), components should still be checking for their target in `noProxy` and say "I don't see my target in `noProxy`, so I guess there is some transparent network routing taking me to a proxy, so I'll attach this additional trusted CA bundle from the Proxy's `trustedCA`..."?

Reply:
if the admin configured a proxy ca for the cluster, you use it on the assumption that they know what they are doing and they put it there for a reason.

yes that means it becomes a valid CA for all connections, that is the power/danger of the proxy CA.  (As consumed by most components today, those CAs will be considered valid even for hosts in no_proxy because they just add the proxy CAs to their bundle, so whether or not their outbound connection goes through the proxy, the CA is in the transport)

@deads2k am i missing something?  do we need to be even more locked down on our use of proxy ca than we are today, such as to ensure that connections to "no proxy" hosts don't utilize the proxy CAs?  I'm sure it would be ideal, it also seems extremely difficult.

Reply:
@wking Why can't we use this enhancement only when network-defined proxies are not available. Any in case of proxy setups CVO going to an upstream Cincy is not an issue. So if Proxy takes the precedence over the endpoint specific CA  it is not a blocker IMO.